mtp_api/apps/transaction/api/cashbook/views.py

```diff
@@ -50,117 +53,151 @@ def get_prison_set_for_user(self):
             return Prison.objects.none()
 
     def get_queryset(self):
-        qs = super(OwnPrisonListModelMixin, self).get_queryset()
-        return qs.filter(prison__in=self.get_prison_set_for_user())
+        return Transaction.objects.filter(prison__in=self.get_prison_set_for_user())
 
 
-class TransactionView(
-    OwnPrisonListModelMixin, mixins.ListModelMixin, viewsets.GenericViewSet,
-):
-    queryset = Transaction.objects.all()
+class GetTransactions(TransactionViewMixin, generics.ListAPIView):
     serializer_class = TransactionSerializer
-    patch_serializer_class = CreditedOnlyTransactionSerializer
-    filter_backends = (
-        filters.DjangoFilterBackend,
-        filters.OrderingFilter,
+    filter_backends = (filters.DjangoFilterBackend,)
+    filter_class = TransactionListFilter
+    action = 'list'
+
+    permission_classes = (
+        IsAuthenticated, CashbookClientIDPermissions,
+        TransactionPermissions
     )
 
-    filter_class = TransactionStatusFilter
 
-    ordering = ('received_at',)
+class CreditTransactions(TransactionViewMixin, generics.GenericAPIView):
+    serializer_class = CreditedOnlyTransactionSerializer
+    action = 'patch_credited'
+
     permission_classes = (
         IsAuthenticated, CashbookClientIDPermissions,
-        IsOwnPrison, TransactionPermissions
+        TransactionPermissions
     )
 
-    def get_queryset(self, filter_by_user=True, filter_by_prison=True):
-        qs = super(TransactionView, self).get_queryset()
+    def get_serializer(self, *args, **kwargs):
+        kwargs['context'] = {
+            'request': self.request,
+            'format': self.format_kwarg,
+            'view': self
+        }
+        return self.serializer_class(*args, **kwargs)
 
-        prison_id = self.kwargs.get('prison_id')
-        user_id = self.kwargs.get('user_id')
+    def patch(self, request, format=None):
+        deserialized = self.get_serializer(data=request.data, many=True)
+        deserialized.is_valid(raise_exception=True)
 
-        if prison_id and filter_by_prison:
-            qs = qs.filter(prison_id=prison_id)
+        transaction_ids = [x['id'] for x in deserialized.data]
+        with transaction.atomic():
+            to_update = self.get_queryset().filter(
+                owner=request.user,
+                pk__in=transaction_ids
+            ).select_for_update()
 
-        if user_id and filter_by_user:
-            qs = qs.filter(owner__id=user_id)
+            ids_to_update = [t.id for t in to_update]
+            conflict_ids = set(transaction_ids) - set(ids_to_update)
 
-        return qs
+            if conflict_ids:
+                return Response(
+                    data={
+                        'error': [
+                            'The following transactions cannot be credited: {ids}'.format(
+                                ids=', '.join(sorted([str(t_id) for t_id in conflict_ids]))
```

Comment:
I think it would be more machine-readable if the conflicted IDs were returned in their own JSON attribute rather than as part of the error message string.